### ISSUE TITLE
Projection of Union types incomplete in nested conditionals

### ISSUE THREAD

**Issue by wence-** [microsoft/pyright#3898]
Consider:

```python
from typing_extensions import reveal_type


def match(x: int | str, y: int | str):
    if (xint := isinstance(x, int)) or isinstance(y, int):
        if xint:
            reveal_type(x) # -> int, correct
            reveal_type(y) # -> int | str, correct
        else:
            reveal_type(x) # -> str, correct
            reveal_type(y) # -> int | str, should be int
```

```
$ pyright --pythonversion 3.10 foo.py
No configuration file found.
No pyproject.toml file found.
stubPath .../typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.269
.../foo.py
  .../foo.py:7:25 - information: Type of "x" is "int"
  .../foo.py:8:25 - information: Type of "y" is "int | str"
  .../foo.py:10:25 - information: Type of "x" is "str"
  .../foo.py:11:25 - information: Type of "y" is "int | str"
0 errors, 0 warnings, 4 informations 
Completed in 0.435sec
```
I expected `reveal_type(y)` in the `else` branch to deduce `int`. The only way we can reach here is if `isinstance(x, int)` is `False` but `isinstance(y, int)` is `True`, so I would expect the union type to projected to `int` (as is deduced in the first branch)

**Comment by erictraut** [comment]
Pyright is working as intended here.

Type narrowing is applied on a per-expression basis. It isn't applied on groups of expressions with conditional dependencies. This is true for all static type checkers (in all languages) that I know of. Although such analysis is theoretically possible, it would be infeasible from a computation and memory perspective.

In your example, `x` is narrowed, but `y` is not conditionally narrowed based on the type of `x`.

You will need to refactor the logic in your code if you want to rely on type narrowing here.

**Comment by wence-** [comment]
Thanks!
